File: python/egg-drop-with-2-eggs-and-n-floors.py
class Solution:
    def get_minimal(n: int, memory) -> int:
        if n in memory:
            return memory[n]
        minimal = n
        for i in range(1, n):
            candidate = max(1 + Solution.get_minimal(n=n - i, memory=memory), i)
            minimal = min(minimal, candidate)
        memory[n] = minimal
        return memory[n]

    def twoEggDrop(self, n: int) -> int:
        return Solution.get_minimal(n=n, memory={0: 0})

File: go/egg-drop-with-2-eggs-and-n-floors.go
func GetMoves(n int, memory map[int]int) int {
	_, ok := memory[n]
	if ok {
		return memory[n]
	}
	minimal := n
	for i := 1; i < n; i++ {
		candidate := GetMoves(n-i, memory) + 1
		if candidate < i {
			candidate = i
		}
		if candidate < minimal {
			minimal = candidate
		}
	}
	memory[n] = minimal
	return memory[n]
}

func twoEggDrop(n int) int {
	memory := make(map[int]int)
	memory[0] = 0
	return GetMoves(n, memory)
}
